File: java/patterns-use-cases/src/main/java/my/example/eventenrichment/PackageTracker.java
package my.example.eventenrichment;

import dev.restate.sdk.Restate;
import dev.restate.sdk.annotation.Handler;
import dev.restate.sdk.annotation.Shared;
import dev.restate.sdk.annotation.VirtualObject;
import dev.restate.sdk.common.StateKey;
import dev.restate.sdk.common.TerminalException;
import dev.restate.sdk.endpoint.Endpoint;
import dev.restate.sdk.http.vertx.RestateHttpServer;
import my.example.eventenrichment.types.LocationUpdate;
import my.example.eventenrichment.types.PackageInfo;

// Package tracking system:
// Digital twin representing a package in delivery with real-time location updates.
// Handlers get called over HTTP or Kafka.
@VirtualObject
public class PackageTracker {

  private static final StateKey<PackageInfo> PACKAGE_INFO =
      StateKey.of("package-info", PackageInfo.class);

  // Called first by the seller over HTTP
  @Handler
  public void registerPackage(PackageInfo packageInfo) {
    // Store the package details in the state
    Restate.state().set(PACKAGE_INFO, packageInfo);
  }

  // Connected to a Kafka topic for real-time location updates
  @Handler
  public void updateLocation(LocationUpdate locationUpdate) {
    var packageInfo =
        Restate.state()
            .get(PACKAGE_INFO)
            .orElseThrow(() -> new TerminalException("Package not found"));

    // Update the package info with the new location
    packageInfo.addLocation(locationUpdate);
    Restate.state().set(PACKAGE_INFO, packageInfo);
  }

  // Called by the delivery dashboard to get the package details
  @Shared
  public PackageInfo getPackageInfo() {
    return Restate.state()
        .get(PACKAGE_INFO)
        .orElseThrow(() -> new TerminalException("Package not found"));
  }

  public static void main(String[] args) {
    RestateHttpServer.listen(Endpoint.bind(new PackageTracker()));
  }
}

// Example API Usage:
/*
curl localhost:8080/PackageTracker/package123/registerPackage -H 'content-type: application/json' -d '{ "finalDestination": "Bridge 6, Amsterdam"}'
curl localhost:8080/PackageTracker/package123/updateLocation -H 'content-type: application/json' -d '{ "timestamp": "2024-12-11T12:00:00Z", "location": "Warehouse A" }'
curl localhost:8080/PackageTracker/package123/getPackageInfo
*/

File: kotlin/patterns-use-cases/src/main/kotlin/my/example/eventenrichment/PackageTracker.kt
package my.example.eventenrichment

import dev.restate.sdk.annotation.Handler
import dev.restate.sdk.annotation.Shared
import dev.restate.sdk.annotation.VirtualObject
import dev.restate.sdk.common.TerminalException
import dev.restate.sdk.http.vertx.RestateHttpServer
import dev.restate.sdk.kotlin.*
import dev.restate.sdk.kotlin.endpoint.endpoint
import kotlinx.serialization.Serializable

@Serializable
data class PackageInfo(
    val finalDestination: String,
    val locations: MutableList<LocationUpdate> = mutableListOf()
)

@Serializable data class LocationUpdate(val timestamp: String, val location: String)

@VirtualObject
class PackageTracker {

  companion object {
    private val PACKAGE_INFO = stateKey<PackageInfo>("package-info")
  }

  @Handler
  suspend fun registerPackage(ctx: ObjectContext, packageInfo: PackageInfo) {
    ctx.set(PACKAGE_INFO, packageInfo)
  }

  @Handler
  suspend fun updateLocation(ctx: ObjectContext, locationUpdate: LocationUpdate) {
    val packageInfo = ctx.get(PACKAGE_INFO) ?: throw TerminalException("Package not found")

    packageInfo.locations.add(locationUpdate)
    ctx.set(PACKAGE_INFO, packageInfo)
  }

  @Shared
  suspend fun getPackageInfo(ctx: SharedObjectContext): PackageInfo {
    return ctx.get(PACKAGE_INFO) ?: throw TerminalException("Package not found")
  }
}

fun main() {
  RestateHttpServer.listen(endpoint { bind(PackageTracker()) })
}
